### PR TITLE
add oracle module

### PR DESCRIPTION
**Pull request by justinbastress**
Added an Oracle module.

Moved `flagsToSet` from mysql into `zgrab2/output.go`; there are probably a few other common output-formatting functions that should go there.

## How to Test

There is no Docker module for this, so there is no integration test.

The Oracle Express Edition can be installed & tested locally.

Oracle does have instructions (https://github.com/oracle/docker-images/tree/master/OracleDatabase) for setting up a container with a end-user-provided Oracle installer, but I ran into issues with this with Docker for Windows.

### CODE REVIEW

**Comment by zakird** [comment]
This looks solid overall. We might want to lowercase some of the output field names to match the other scanners.

**Comment by zakird** [comment]
Do you have a sense of what versions of oracle this is compatible with?

**Comment by justinbastress** [comment]
For output fields, you mean the flags names / constants, e.g. `oracle.data.result.handshake_log.global_service_options = { "HANG_ON": true, "URGENT_IO": true }` should be converted to `oracle.data.result.handshake_log.global_service_options = { "hang_on": true, "urgent_io": true }`?

The thinking was that the values would match e.g. the names from C-header includes or docs, which are typically `C_DEFINE_MACRO_STYLE`. In this particular case, since we have no headers / official docs, this isn't a constraint here, but I went with the uppercase to make it consistent with the situations where we do have docs.

But, I don't have strong feelings either way.

**Comment by justinbastress** [comment]
Unsurprisingly the 1% zmap scan only turned up 333 results with 1521 open; running those through zgrab2, only 7 returned any data, and two were apparently bogus (one looks like MySQL and the other SSH).

The other 5 were positively detected, and we got version numbers for each (though in different places -- 4 were "refused" packets, so the version was pulled from the error, while the other pulled the version number from the "accepted" location).

The error cases look to be legit, since the error code (12504) corresponds to "No SID given", which is correct -- the default connection descriptor gives no SID, which in the 5th case (and in my default install of Oracle Express Edition) allows the server to pick a default SID.

**Comment by justinbastress** [comment]
Sample output:
```
[
  {
    "data": {
      "oracle": {
        "result": {
          "handshake": {
            "accept_version": 0,
            "did_resend": false,
            "refuse_error": [
              {
                "key": "DESCRIPTION.TMP",
                "value": ""
              },
              {
                "key": "DESCRIPTION.VSNNUM",
                "value": "169870336"
              },
              {
                "key": "DESCRIPTION.ERR",
                "value": "12504"
              },
              {
                "key": "DESCRIPTION.ERROR_STACK.ERROR.CODE",
                "value": "12504"
              },
              {
                "key": "DESCRIPTION.ERROR_STACK.ERROR.EMFI",
                "value": "4"
              }
            ],
            "refuse_error_raw": "(DESCRIPTION=(TMP=)(VSNNUM=169870336)(ERR=12504)(ERROR_STACK=(ERROR=(CODE=12504)(EMFI=4))))",
            "refuse_reason_app": "0x22",
            "refuse_reason_sys": "0x00",
            "refuse_version": "10.2.0.4.0"
          }
        },
        "status": "success",
        "timestamp": "2018-03-02T12:02:10-05:00"
      }
    },
    "ip": "88.255.66.78"
  },
  {
    "data": {
      "oracle": {
        "result": {
          "handshake": {
            "accept_version": 0,
            "did_resend": false,
            "refuse_error": [
              {
                "key": "DESCRIPTION.TMP",
                "value": ""
              },
              {
                "key": "DESCRIPTION.VSNNUM",
                "value": "186646784"
              },
              {
                "key": "DESCRIPTION.ERR",
                "value": "12504"
              },
              {
                "key": "DESCRIPTION.ERROR_STACK.ERROR.CODE",
                "value": "12504"
              },
              {
                "key": "DESCRIPTION.ERROR_STACK.ERROR.EMFI",
                "value": "4"
              },
              {
                "key": "DESCRIPTION.ERROR_STACK.ERROR.CODE",
                "value": "12504"
              },
              {
                "key": "DESCRIPTION.ERROR_STACK.ERROR.EMFI",
                "value": "4"
              }
            ],
            "refuse_error_raw": "(DESCRIPTION=(TMP=)(VSNNUM=186646784)(ERR=12504)(ERROR_STACK=(ERROR=(CODE=12504)(EMFI=4))(ERROR=(CODE=12504)(EMFI=4))))",
            "refuse_reason_app": "0x22",
            "refuse_reason_sys": "0x00",
            "refuse_version": "11.2.0.1.0"
          }
        },
        "status": "success",
        "timestamp": "2018-03-02T12:02:10-05:00"
      }
    },
    "ip": "183.96.28.124"
  },
  {
    "data": {
      "oracle": {
        "error": "server returned invalid data",
        "result": null,
        "status": "unknown-error",
        "timestamp": "2018-03-02T12:02:10-05:00"
      }
    },
    "ip": "219.105.235.162"
  },
  {
    "data": {
      "oracle": {
        "result": {
          "handshake": {
            "accept_version": 312,
            "connect_flags0": {
              "SERVICES_WANTED": true,
              "UNKNOWN_20": true,
              "UNKNOWN_40": true
            },
            "connect_flags1": {
              "SERVICES_WANTED": true,
              "UNKNOWN_40": true
            },
            "did_resend": true,
            "global_service_options": {
              "HEADER_CHECKSUM": true,
              "UNKNOWN_0001": true,
              "UNKNOWN_0040": true
            },
            "nsn_service_versions": {
              "Authentication": "11.2.0.2.0",
              "DataIntegrity": "11.2.0.2.0",
              "Encryption": "11.2.0.2.0",
              "Supervisor": "11.2.0.2.0"
            }
          }
        },
        "status": "success",
        "timestamp": "2018-03-02T12:02:10-05:00"
      }
    },
    "ip": "122.99.223.18"
  }
]
```

File: modules/oracle/scanner.go [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+// Default Port: 1521 (TCP)
+//
+// The scan does the first part of a TNS handshake, prior to the point where
+// any actual authentication is required; the happy case goes
+//  ->  Connect(--client-version, --min-server-version, --connect-descriptor)
+// <-   Resend
+//  ->  Connect(exact same data)
+// <-   Accept(server_version)
+//  ->  Data: Native Service Negotiation
+// <-   Data: Native Service Negotiation(component release versions)
```
Comment (dadrian) [inline modules/oracle/scanner.go]: This probably renders poorly on godoc

File: schemas/oracle.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+from zschema.leaves import *
+from zschema.compounds import *
+import zschema.registry
+
+import schemas.zcrypto as zcrypto
+import schemas.zgrab2 as zgrab2
+
+# Get a dict of all keys, mapping key -> true 
+def flagsSet(keys):
+    return SubRecord(dict((key, Boolean()) for key in keys))
```
Comment (dadrian) [inline schemas/oracle.py]: Arguably more Pythonic to write `SubRecord({key: Boolean() for key in keys})`, but this doesn't really matter.

File: schemas/oracle.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    "INTERCHANGE_INVOLVED",
+    "SERVICES_ENABLED",
+    "SERVICES_LINKED_IN",
+    "SERVICES_REQUIRED",
+    "UNKNOWN_20",
+    "UNKNOWN_40",
+    "UNKNOWN_80",
+]
+
+nsn_services = [
```
Comment (dadrian) [inline schemas/oracle.py]: I think this is unused

**Pull request merged**